Add NetMHC 4.0 to mhctools

tavinathanson opened the pull request:
No more `NetMHC` class.

Confusingly, test code relies on `netMHC` = 4.0 and `netMHC-3.4` = 3.4. Happy to consider less confusing alternatives.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/mhctools/48)

<!-- Reviewable:end -->

tavinathanson committed:
Add NetMHC 4.0

tavinathanson committed:
Fix merge conflict residue

tavinathanson committed:
Add version test

iskandr commented on README.md:
```diff
@@ -28,7 +28,8 @@ strongest_predicted_binder = epitope_collection[0]
 
 The following models are available in `mhctools`:
-* `NetMHC`: requires locally installed version of [NetMHC](http://www.cbs.dtu.dk/services/NetMHC/)
+* `NetMHC3`: requires locally installed version of [NetMHC 3.x](http://www.cbs.dtu.dk/services/NetMHC/)
```
This should probably link to: http://www.cbs.dtu.dk/services/NetMHC-3.4/

iskandr commented on mhctools/file_formats.py:
```diff
@@ -91,13 +91,13 @@ def split_stdout_lines(stdout):
         sequence_key_mapping=None):
     """
-    Parse the output format for NetMHC, which looks like:
+    Parse the output format for NetMHC 3.x, which looks like:
```
Man, I'm sorry that had to be added. Maybe this is a form of Danish humor?

iskandr commented:
Other than the link, looks good to me. 

I guess now I have to fix Topiary to use NetMHC4

tavinathanson committed:
Update prediction method name to be versioned

tavinathanson committed:
Add a test for NetMHC 4.0 output parsing

tavinathanson commented:
@iskandr two new additions:
- I changed the `prediction_method_name` from `netmhc` to `netmhc3` and `netmhc4`; any reason this would break things?
- I added a parser test.

tavinathanson committed:
Update README link

iskandr commented:
@tavinathanson 

I think the `prediction_method_name` change should be OK. :pray: 

LGTM (modulo tests)

JPFinnigan commented:
Hey, 

I think the change from `netmhc` to `netmhc3` and `netmhc4` may have broken `varlens-variants`. I'll file this as a separate issue in that repository.

tavinathanson commented:
@JPFinnigan Yikes, it's not clear to me why it would have, but I'll look into it!

JPFinnigan commented:
Hey @tavinathanson,

Here's what I experienced. Tried to run varlens-variants w/:  

```
JPF-MBP:varlens johnfinnigan$ varlens-variants \
> --include-read-evidence \
> --reads ~/Desktop/TEMP/Results/RNA/Tumor_B16.F10/ISMMS/Tumor_B16.F10_0810.127A/BAM/HISAT2/Tumor_B16.F10_0810.127A.HISAT2.sorted.bam \
> --reads ~/Desktop/TEMP/Results/RNA/Tumor_B16.F10/ISMMS/Tumor_B16.F10_0810.131B/BAM/HISAT2/Tumor_B16.F10_0810.131B.HISAT2.sorted.bam \
> --variants ~/Desktop/TEMP/Results/WES/Tumor_B16.F10_0810/ISMMS/VCF/MuTect/Tumor_B16_F10_0810.mutect.targets.pass.vcf \
> --variants ~/Desktop/TEMP/Results/WES/Tumor_B16.F10_0810/ISMMS/VCF/Strelka/results/passed.somatic.snvs.vcf \
> --out ~/Desktop/Tumor.B16.F10.0810.mutect.targets.pass.vcf_Tumor.B16.F10.0810.strelka.passed.snvs.vcf_Tumor.B16.F10.0810.127A.HISAT2.sorted.bam_Tumor.B16.F10.0810.131B.HISAT2.sorted.bam_varlens.csv
```

Received the following error message:

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/bin/varlens-variants", line 9, in <module>
    load_entry_point('varlens', 'console_scripts', 'varlens-variants')()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pkg_resources/__init__.py", line 547, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2720, in load_entry_point
    return ep.load()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2380, in load
    return self.resolve()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2386, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/Users/johnfinnigan/Desktop/Utilities/Varlens/varlens/varlens/commands/variants.py", line 22, in <module>
    from .. import variant_includes
  File "/Users/johnfinnigan/Desktop/Utilities/Varlens/varlens/varlens/variant_includes.py", line 12, in <module>
    from . import mhc_binding
  File "/Users/johnfinnigan/Desktop/Utilities/Varlens/varlens/varlens/mhc_binding.py", line 19, in <module>
    import topiary
  File "/Users/johnfinnigan/Desktop/Utilities/Topiary/topiary/topiary/__init__.py", line 7, in <module>
    from .predict_epitopes import (
  File "/Users/johnfinnigan/Desktop/Utilities/Topiary/topiary/topiary/predict_epitopes.py", line 20, in <module>
    from .commandline_args import (
  File "/Users/johnfinnigan/Desktop/Utilities/Topiary/topiary/topiary/commandline_args.py", line 185, in <module>
    "netmhc": mhctools.NetMHC,
AttributeError: 'module' object has no attribute 'NetMHC'
```

I'm thinking this means that varlens is looking for, but not finding NetMHC. Presumably, this is because mhctools was just updated to go from `NetMHC` to `NetMHC3` `NetMHC4`. Perhaps not?

tavinathanson commented:
Ah, yeah: the issue is that `topiary` is broken (see https://github.com/hammerlab/topiary/issues/44), and `varlens` uses `topiary`. I'll put out either a fix or revert today. I'll defer the `varlens` question to @timodonnell.

JPFinnigan commented:
Thanks @tavinathanson. Sorry about that, I didn't mean to bother you guys on a Sunday afternoon. 

Take care!

tavinathanson commented:
No worries, I didn't mean to break software you were using :)

tavinathanson commented:
@JPFinnigan I think your problem should be resolved if you use the very latest version of `mhctools` (2.1); I just introduced `NetMHC`, which automatically determines whether the software is NetMHC 3 or 4, and that should hopefully fix that `varlens`/`topiary` issue.
